packages/@aws-cdk/custom-resources/lib/provider-framework/runtime/cfn-response.ts

```diff
@@ -50,11 +53,41 @@ export let includeStackTraces = true; // for unit tests
 
 export function safeHandler(block: (event: any) => Promise<void>) {
   return async (event: any) => {
+
+    // ignore DELETE event when the physical resource ID is the marker that
+    // indicates that this DELETE is a subsequent DELETE to a failed CREATE
+    // operation.
+    if (event.RequestType === 'Delete' && event.PhysicalResourceId === CREATE_FAILED_PHYSICAL_ID_MARKER) {
+      log(`ignoring DELETE event caused by a failed CREATE event`);
+      await submitResponse('SUCCESS', event);
+      return;
+    }
+
     try {
       await block(event);
     } catch (e) {
       // tell waiter state machine to retry
-      if (e instanceof Retry) { throw e; }
+      if (e instanceof Retry) {
+        log(`retry requested by handler`);
+        throw e;
+      }
+
+      if (!event.PhysicalResourceId) {
+        // special case: if CREATE fails, which usually implies, we usually don't
+        // have a physical resource id. in this case, the subsequent DELETE
+        // operation does not have any meaning, and will likely fail as well. to
+        // address this, we use a marker so the provider framework can simply
+        // ignore the subsequent DELETE.
+        if (event.RequestType === 'Create') {
+          log(`CREATE failed, responding with a marker physical resource id so that the subsequent DELETE will be ignored`);
+          event.PhysicalResourceId = CREATE_FAILED_PHYSICAL_ID_MARKER;
```

Comment:
Since we're making code changes anyway, I don't understand why we simply couldn't extend the protocol between the framework and the handler to properly indicate whether a resource got created or not.

My concern is a resource that requires multi-step initialization (create, update, update, ...) is now going to leak resources if an update step fails.

Is it really that much more work to do it properly right away, rather than put a hack in place and do it properly later (if ever)?

Reply:
I disagree that this is a hack. I’d argue that this is the right behavior for 95% of the cases, and I’ve created the referenced issue for people to tell me otherwise.

Reply:
I also think that this behavior will not change. If the user handler throws an exception during CREATE (which is currently the only way to signal a failure) without any special instructions, I'd expect this to be the behavior (i.e. ignore the subsequent DELETE). If, in the future, we want to allow people to customize PhysicalResourceId for failures (please read https://github.com/aws/aws-cdk/issues/5524), it will be supported for any type of failure and will apply to the CREATE case as well.